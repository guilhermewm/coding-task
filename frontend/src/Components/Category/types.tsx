export type Category = {
    name: string,
    keywords: string[]
}

export type CategoryProps = {
    category: Category
}