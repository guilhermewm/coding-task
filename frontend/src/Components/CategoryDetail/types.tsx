export type CategoryDetailProps = {
    categoryName: string;
};