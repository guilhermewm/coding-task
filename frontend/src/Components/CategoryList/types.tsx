import { Category } from "../Category/types";

export type CategoryListProps = {
  categories: Category[];
  onChangeCategory: Function;
};