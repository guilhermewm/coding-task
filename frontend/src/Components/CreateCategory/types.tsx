import { Category } from "../Category/types";

export type CreateCategoryProps = {
    createCategory: (name: string) => void,
    error: string,
    success: string,
    loading: boolean
};