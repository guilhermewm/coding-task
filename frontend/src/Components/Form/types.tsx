export type FormProps = {
    type: string,
    submit: Function,
    error: string,
    success: string,
    loading?: boolean
}
