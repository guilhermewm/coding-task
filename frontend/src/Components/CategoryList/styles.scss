@import '../Styles/colors.scss';

.category-list {
  list-style: none;
  margin: 0;
  padding: 0;
  height: calc(100vh - 7vh - 20px);
  overflow-y: auto;
  -webkit-overflow-scrolling: auto;  
  min-width: 275px;

  display: flex;
  flex-direction: column;

  li {
    border-top: solid 1px #888888;
    background-color: $base-color;
  }

  li:last-child {
    border-bottom: solid 1px #888888;
  }

  li:nth-child(2n + 1) {
    background-color: $light-secondary-color;
  }

  li:hover {
    background-color: $dark-primary-color;
    color: $base-color;
  }
}