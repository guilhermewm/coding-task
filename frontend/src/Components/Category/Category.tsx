import React, { FC } from "react";
import { CategoryProps } from "./types";
import './styles.scss';
import { Link } from "react-router-dom";

const CategoryCard: FC<CategoryProps> = ({category}) => {
  return (
    
    <Link to={{search: `?name=${category.name}`}}> 
      <div className="category">
        <strong>{category.name}</strong>
        <small>{category.keywords.join(', ')}</small>
      </div>
    </Link>
  
  );
};

export default CategoryCard;
