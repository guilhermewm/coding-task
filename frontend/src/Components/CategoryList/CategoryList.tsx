import React, { FC } from "react";
import Category from "../Category/Category";
import Header from "../Header/Header";
import { CategoryListProps } from "./types";
import { Link } from "react-router-dom";

import './styles.scss';

const CategoryList: FC<CategoryListProps> = ({ categories, changeCategory }) => {
  return (
    <>
      <Header>
        <h1>Categories</h1>
        <Link to={{search: ''}}> 
          <svg className="add" onClick={() => changeCategory({})}
            xmlns="http://www.w3.org/2000/svg"
            width="50"
            height="50"
            viewBox="0 0 52 52"
          >
            <path d="M26 0C11.664 0 0 11.663 0 26s11.664 26 26 26 26-11.663 26-26S40.336 0 26 0zm0 50C12.767 50 2 39.233 2 26S12.767 2 26 2s24 10.767 24 24-10.767 24-24 24z" />
            <path d="M38.5 25H27V14c0-.553-.448-1-1-1s-1 .447-1 1v11H13.5c-.552 0-1 .447-1 1s.448 1 1 1H25v12c0 .553.448 1 1 1s1-.447 1-1V27h11.5c.552 0 1-.447 1-1s-.448-1-1-1z" />
          </svg>   
        </Link>
      </Header>
       {!!categories.length && (
         <ul className="category-list">
          {categories?.map((category, index) => (
            <li key={index} onClick={() => changeCategory(category)}>
              <Category category={category} />
            </li>
          ))}
         </ul>
       )}        
    </>
  );
};

export default CategoryList;
