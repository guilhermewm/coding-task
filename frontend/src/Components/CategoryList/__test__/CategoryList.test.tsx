 // Link.react.test.js
import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, render } from '@testing-library/react';

import { Category } from "../../Category/types";
import { BrowserRouter as Router } from "react-router-dom";
import CategoryList from "../CategoryList";

describe("CategoryList", () => {
  describe("Snapshot", () => {
    it("Should render and empty list", () => {
      const component = renderer.create(
        <Router>
          <CategoryList onChangeCategory={() => {}} categories={[]} />
        </Router>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("Should render the list correctly", () => {
      const categories: Category[] = [
        {
          name: "cars",
          keywords: ["audi", "ferrari", "bmw"],
        },
        {
          name: "music",
          keywords: ["guitar", "drums", "bass"],
        },
      ];
      const component = renderer.create(
        <Router>
          <CategoryList onChangeCategory={() => {}} categories={categories} />
        </Router>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("Behavior", () => {
    it("Should call the onChange function when the button is clicked", () => {
      const categories: Category[] = [
        {
          name: "cars",
          keywords: ["audi", "ferrari", "bmw"],
        },
        {
          name: "music",
          keywords: ["guitar", "drums", "bass"],
        },
      ];

      const changeCategory = jest.fn();

      const component = render(
        <Router>
          <CategoryList onChangeCategory={changeCategory} categories={categories} />
        </Router>
      );

      const firstItem = component.container.getElementsByTagName('li')[0]
      fireEvent.click(firstItem);
      expect(changeCategory).toHaveBeenCalledWith(categories[0]);
    });
  });
});
