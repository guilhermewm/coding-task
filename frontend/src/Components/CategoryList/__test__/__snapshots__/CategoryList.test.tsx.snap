// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryList Snapshot Should render and empty list 1`] = `
<div
  className="header"
>
  <h1>
    Categories
  </h1>
  <a
    href="/"
    onClick={[Function]}
  >
    <svg
      className="add"
      height="50"
      onClick={[Function]}
      viewBox="0 0 52 52"
      width="50"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26 0C11.664 0 0 11.663 0 26s11.664 26 26 26 26-11.663 26-26S40.336 0 26 0zm0 50C12.767 50 2 39.233 2 26S12.767 2 26 2s24 10.767 24 24-10.767 24-24 24z"
      />
      <path
        d="M38.5 25H27V14c0-.553-.448-1-1-1s-1 .447-1 1v11H13.5c-.552 0-1 .447-1 1s.448 1 1 1H25v12c0 .553.448 1 1 1s1-.447 1-1V27h11.5c.552 0 1-.447 1-1s-.448-1-1-1z"
      />
    </svg>
  </a>
</div>
`;

exports[`CategoryList Snapshot Should render the list correctly 1`] = `
Array [
  <div
    className="header"
  >
    <h1>
      Categories
    </h1>
    <a
      href="/"
      onClick={[Function]}
    >
      <svg
        className="add"
        height="50"
        onClick={[Function]}
        viewBox="0 0 52 52"
        width="50"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26 0C11.664 0 0 11.663 0 26s11.664 26 26 26 26-11.663 26-26S40.336 0 26 0zm0 50C12.767 50 2 39.233 2 26S12.767 2 26 2s24 10.767 24 24-10.767 24-24 24z"
        />
        <path
          d="M38.5 25H27V14c0-.553-.448-1-1-1s-1 .447-1 1v11H13.5c-.552 0-1 .447-1 1s.448 1 1 1H25v12c0 .553.448 1 1 1s1-.447 1-1V27h11.5c.552 0 1-.447 1-1s-.448-1-1-1z"
        />
      </svg>
    </a>
  </div>,
  <ul
    className="category-list"
  >
    <li
      onClick={[Function]}
    >
      <a
        href="/?name=cars"
        onClick={[Function]}
      >
        <div
          className="category"
        >
          <strong>
            cars
          </strong>
          <small>
            audi, ferrari, bmw
          </small>
        </div>
      </a>
    </li>
    <li
      onClick={[Function]}
    >
      <a
        href="/?name=music"
        onClick={[Function]}
      >
        <div
          className="category"
        >
          <strong>
            music
          </strong>
          <small>
            guitar, drums, bass
          </small>
        </div>
      </a>
    </li>
  </ul>,
]
`;
